pkg/router/template/router.go

```diff
@@ -873,72 +873,103 @@ type endpointsCounter func(key string) int32
 // getServiceUnits returns a map of service keys to their weights.
 // The requests are loadbalanced among the services referenced by the route.
 // The weight (0-256, default 1) sets the relative proportions each
-// service gets (weight/sum_of_weights fraction of the requests).
+// service gets (weight/sum_of_weights) fraction of the requests.
+// Default when service weight is omitted is 1.
 // When the weight is 0 no traffic goes to the service. If they are
 // all 0 the request is returned with 503 response.
 // For each service, the requests are distributed among the endpoints.
 // Each endpoint gets weight/numberOfEndpoints portion of the requests.
+// The largest weight per endpoint is scaled to 256 to permit better
+// percision results.
+// returns weight per endpoint for each service
 // The above assumes roundRobin scheduling.
 func getServiceUnits(counter endpointsCounter, route *routeapi.Route) map[string]int32 {
```

Comment:
can you provide a full table test that shows all the pathological edge cases?  Hard to read, would rather review inputs and outputs

Reply:
I'll add more. Its basically the range 1-256 has limited precision. The previous version gave away some of that by using integers and distributing the actual scaled service weight. This version goes with floating point and scales the weight per endpoint. When the scaled weight per endpoint is less than 1.0, we end up setting all endpoints in the service to 1. This can skew the desired distribution. Also, the previous version didn't handle very large services (more than 256 endpoints) e.g., 500 endpoints in "A" and 500 endpoints in "B". In this case for "A" 10, "B" 20 the new algorithm will set all "A" endpoints to 128 and all "B" endpoints to 256 which will yield the desired distribution. Previously, all endpoints would be set to 1 which is a 50-50 split.

Reply:
No, I mean a unit table test that correctly validates the edge cases here.  Each input (set of services and weights) and output (adjusted weights).  I.e. full table driven unit test.

Reply:
Working on this

Reply:
@smarterclayton The edge case is when the weight per endpoint (scaled) is less than 1 since the minimum weight per endpoint is 1.